src/mlpack/methods/ann/activation_functions/hard_sigmoid_function.hpp

```diff
@@ -0,0 +1,101 @@
+/**
+ * @file hard_sigmoid_function.hpp
+ * @author Bishwa Karki
+ *
+ * Definition and implementation of the hard sigmoid function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_HARD_SIGMOID_FUNCTION_HPP
+#define MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_HARD_SIGMOID_FUNCTION_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <algorithm>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The hard sigmoid function, defined by
+ *
+ * @f{eqnarray*}{
+ * if x < -2.5 \\
+ *    @return 0
+ * elseif x > 2.5 \\
+ *    @return 1
+ * else \\
+ *    @return x * .2 + .5
+ * endif
+ * if x > 2.5 || x < -2.5
+ *    f'(x) &=& 0 \\
+ * else
+ *    f'(x) &=& 0.2
+ * @f}
+ */
+class HardSigmoidFunction
+{
+ public:
+  /**
+   * Computes the hard sigmoid function.
+   *
+   * @param x Input data.
+   * @return f(x).
+   */
+  static double Fn(const double x)
+  {
+    return std::min(1.0, std::max(0.0, 0.2 * x + 0.5));
+  }
+
+  /**
+   * Computes the hard sigmoid function.
+   *
+   * @param x Input data.
+   * @param y The resulting output activation.
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType& x, OutputVecType& y)
+  {
+    y.set_size(size(x));
+
+    for (size_t i = 0; i < x.n_elem; i++)
+      y[i] = Fn(x(i));
+  }
+
+  /**
+   * Computes the first derivatives of hard sigmoid function.
+   *
+   * @param y Input data.
+   * @return f'(x)
+   */
+  static double Deriv(const double y)
+  {
+    if (y < -2.5 || y > 2.5)
+    {
+      return 0.0;
+    }
+    return 0.2;
+  }
+
+  /**
+   * Computes the first derivatives of the hard sigmoid function.
+   *
+   * @param y Input activations.
+   * @param x The resulting derivatives.
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Deriv(const InputVecType& y, OutputVecType& x)
+  {
+    for (size_t i = 0; i < y.n_elem; i++)
```

Comment:
You forgot the `x.set_size(size(y));`. Also, it should be `x(i) = Deriv(y(i));` inside the loop.

Reply:
@ShikharJ Thank you :)